packages/react/menu/src/Menu.tsx

```diff
@@ -621,6 +636,206 @@ const MenuItemIndicator = React.forwardRef((props, forwardedRef) => {
 
 MenuItemIndicator.displayName = ITEM_INDICATOR_NAME;
 
+/* -------------------------------------------------------------------------------------------------
+ * MenuSubMenu
+ * -----------------------------------------------------------------------------------------------*/
+
+const SUB_MENU_NAME = 'MenuSubMenu';
+
+type MenuSubMenuContextValue = {
+  triggerRef: React.RefObject<HTMLDivElement>;
+  contentId: string;
+  focusFirstItem: boolean;
+  onMouseOpen(): void;
+  onKeyboardOpen(): void;
+};
+
+const [SubMenuProvider, useSubMenuContext] = createContext<MenuSubMenuContextValue>(SUB_MENU_NAME);
+
+type MenuSubMenuOwnProps = {
+  open?: boolean;
+  onOpenChange?(open: boolean): void;
+  defaultOpen?: boolean;
+};
+
+const MenuSubMenu: React.FC<MenuSubMenuOwnProps> = (props) => {
```

Comment:
Adhering to our naming convention though it reads a bit weird. Any thoughts on something better?

Reply:
Haha yeah this does read a bit weird but tbh, I'm thinking it would be better DX not to expose these parts to consumers and then in that case the naming doesn't matter too much.

My thinking is that we would render these conditionally internally based on whether the menu is nested inside a menu in the React Tree... something like:

```jsx
const IsSubMenuContext = React.createContext(false);

const Menu = React.forwardRef((props, forwardedRef) => {
  const isSubMenu = React.useContext(IsSubMenuContext);
  
  if (isSubMenu) {
    return <MenuSubMenu {...props} ref={forwardedRef} />
  } else {
    return (
		<IsSubMenuContext.Provider value={true}>
			<MenuImpl {...props} ref={forwardedRef} />
		</IsSubMenuContext.Provider>
	);
  }
});

const MenuSubMenu = React.forwardRef((props, forwardedRef) => {
  // any logic specific to submenu and then
  return <MenuImpl {...props} ref={forwardedRef} />
});

const MenuImpl = React.forwardRef((props, forwardedRef) => {
  // common logic across menu and submenu types
  return <Primitive {...props} ref={forwardedRef} />
});
```

And the same for the other parts. That way the consumers can just nest `Menu`s inside `Menu`s.

Reply:
We discussed this on a call but for posterity – Totally agree 💯 , this is the way I would expect it to work as well. Originally wasn't sure on the user expectation hence the explicitness but our discussion cleared up doubts.